AX: Move rowHeader to AXCoreObject

hoffmanjoshua opened the pull request:
#### 7c604faeacdeac530781ab88889531783b830279
<pre>
AX: Move rowHeader to AXCoreObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=293129">https://bugs.webkit.org/show_bug.cgi?id=293129</a>
<a href="https://rdar.apple.com/problem/151475547">rdar://problem/151475547</a>

Reviewed by Tyler Wilcock.

We can move rowHeader to a shared implementation in AXCoreObject, now
that we cache the right element names.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::rowHeader):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::rowHeader): Deleted.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp:
(WebCore::AccessibilityARIAGridRow::rowHeader): Deleted.
* Source/WebCore/accessibility/AccessibilityARIAGridRow.h:
(isType):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityARIAGridRowInstance const):
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::rowHeader): Deleted.
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/295069@main">https://commits.webkit.org/295069@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02198c258f35d1056b5fa2f7ef4bac213f5d9db8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103886 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23589 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109079 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54660 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23942 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32134 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78921 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93707 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59250 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18474 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11758 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53914 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88206 "Found 1 new API test failure: TestWebKitAPI.CARingBufferTest.FetchTimeBoundsConsistent (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111466 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22883 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87930 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31404 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89908 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87582 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22314 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32471 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10197 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25550 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30971 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36281 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30764 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34101 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32325 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
```diff
@@ -490,7 +489,7 @@ class AXIsolatedObject final : public AXCoreObject {
     bool isAccessibilityRenderObject() const final;
     bool isAccessibilityTableInstance() const final;
-    bool isAccessibilityARIAGridRowInstance() const final { return false; }
+    bool isAccessibilityARIAGridRowInstance() const final { return boolAttributeValue(AXProperty::IsARIAGridRow) || isARIATreeGridRow(); }
```
We should rename this function, as it's misleading in its current state. `AXIsolatedObject` is never an "instance" of `AccessibilityARIAGridRow` (it's an instance of `AXIsolatedObject`). And that's especially important when you consider these "is-instance-of" functions are used to implement casting via `is<T>` implementations — a mistake can cause a security vulnerability.

twilco replied:
Ideally we could make `isAccessibilityARIAGridRowInstance` virtual on `AccessibilityObject`, not on `AXCoreObject`, then name this function something else that actually does make sense for both AXIsolatedObjects, and live objects

hoffmanjoshua replied:
Fixed! Made the AXCoreObject method named `isARIAGridRow`, which matches the property name, and then moved this to be virtual on AccessibilityObject.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -107,7 +107,7 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
     bool isAccessibilityTableInstance() const override { return false; }
     virtual bool isAccessibilityTableColumnInstance() const { return false; }
-    bool isAccessibilityARIAGridRowInstance() const override { return false; }
+    virtual bool isAccessibilityARIAGridRowInstance() const { return false; };
```
I don't think we need the semicolon after the curly brace

hoffmanjoshua committed:
AX: Move rowHeader to AXCoreObject
https://bugs.webkit.org/show_bug.cgi?id=293129
rdar://problem/151475547

Reviewed by Tyler Wilcock.

We can move rowHeader to a shared implementation in AXCoreObject, now
that we cache the right element names.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::rowHeader):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::rowHeader): Deleted.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp:
(WebCore::AccessibilityARIAGridRow::rowHeader): Deleted.
* Source/WebCore/accessibility/AccessibilityARIAGridRow.h:
(isType):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityARIAGridRowInstance const):
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::rowHeader): Deleted.
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/295069@main
